### PR TITLE
Changing auth documentation for C++.

### PR DESCRIPTION
**Pull request by jboeuf**
There's been quite a lot of changes in the C++ API since this document was last updated.

Assigning to @yang-g now to check the C++ sample code but @a11r and @LisaFC will have the final say on the text.

### CODE REVIEW

**Comment by jboeuf** [comment]
@jtattermusch please feel free to review as well.

**Comment by jboeuf** [comment]
@yang-g do you think that you can have a look soonish? Thanks!

File: docs/guides/auth.md [hunk 1]
```diff
@@ -22,37 +22,50 @@ authentication.
...
+
 
 ###SSL/TLS for server authentication and encryption
 
 This is the simplest authentication scenario, where a client just wants to
 authenticate the server and encrypt all data.
 
-```
+```cpp
 SslCredentialsOptions ssl_opts;  // Options to override SSL params, empty by default
```
Comment (yang-g) [inline docs/guides/auth.md]: The two comments seem wrong here.

File: docs/guides/auth.md [hunk 2]
```diff
@@ -66,45 +79,75 @@ passed to the factory method.
...
 ###Authenticating with Google
 
-gRPC applications can use a simple API to create a credential that works in various deployment scenarios.
+gRPC applications can use a simple API to create a credential that works in
+various deployment scenarios.
 
-```
-std::unique_ptr<Credentials> creds = CredentialsFactory::GoogleDefaultCredentials();
+```cpp
+auto creds = CredentialsFactory::GoogleDefaultCredentials();
```
Comment (yang-g) [inline docs/guides/auth.md]: Remove `CredentialsFactory::`.

File: docs/guides/auth.md [hunk 3]
```diff
@@ -66,45 +79,75 @@ passed to the factory method.
...
+    metadata->insert(std::make_pair("x-custom-auth-ticket", ticket_));
+    return Status::OK;
+  }
+
+ private:
+  grpc::string ticket_;
+};
+
+auto call_creds =
+    MetadataCredentialsFromPlugin(std::unique_ptr<MetadataCredentialsPlugin>(
```
Comment (yang-g) [inline docs/guides/auth.md]: Some items do not have `grpc::`, such as Status, AuthContext, MetadataCredentialsPlugin.

**Comment by yang-g** [comment]
C++ code example LGTM except for the minor issues.

**Comment by jboeuf** [comment]
Thanks much. I addressed the comments. I also used grpc:: for all grpc functions for the sake of clarity.

**Comment by yang-g** [comment]
LGTM

File: docs/guides/auth.md [hunk 4]
```diff
@@ -22,38 +22,50 @@ authentication.
...
+- Channel Credentials which are attached to a channel such as SSL credentials.
+- Call Credentials which are attached to a call (or `ClientContext` in C++).
+
+Credentials can be composed using `CompositeChannelCredentials` which associates
+a `ChannelCredentials` and a `CallCredentials` in order to create a new
+`ChannelCredentials`. The result will, for each call on the channel, send the
+authentication data associated with the composed `CallCredentials`.
+
+For example, a `ChannelCredentials` could be created from an `SslCredentials`
+and an `AccessTokenCredentials`. The result, applied to a `Channel` would send
```
Comment (LisaFC) [inline docs/guides/auth.md]: minor change: The result when applied to a `Channel` would send the appropriate access token for each call on this channel

File: docs/guides/auth.md [hunk 5]
```diff
@@ -22,38 +22,50 @@ authentication.
...
+Credentials objects.
+
+Credentials can be of two types:
+
+- Channel Credentials which are attached to a channel such as SSL credentials.
+- Call Credentials which are attached to a call (or `ClientContext` in C++).
+
+Credentials can be composed using `CompositeChannelCredentials` which associates
+a `ChannelCredentials` and a `CallCredentials` in order to create a new
+`ChannelCredentials`. The result will, for each call on the channel, send the
```
Comment (LisaFC) [inline docs/guides/auth.md]: Slightly long sentence, how about:
Credentials can be composed using `CompositeChannelCredentials`. A `CompositeChannelCredentials` associates a `ChannelCredentials` and a `CallCredentials` to create a new `ChannelCredentials`.

File: docs/guides/auth.md [hunk 6]
```diff
@@ -22,38 +22,50 @@ authentication.
...
-credentials (e.g., SSL client certs or Service Account Keys), and use the
-credentials while creating a gRPC channel to any server. Depending on the type of
-credential supplied, the channel uses the credentials during the initial SSL/TLS
-handshake with the server, or uses  the credential to generate and attach Access
-Tokens to each request being made on the channel.
+Credentials objects.
+
+Credentials can be of two types:
+
+- Channel Credentials which are attached to a channel such as SSL credentials.
```
Comment (LisaFC) [inline docs/guides/auth.md]: *Channel credentials*, which are attached to a channel, such as SSL credentials

File: docs/guides/auth.md [hunk 7]
```diff
@@ -22,38 +22,50 @@ authentication.
...
+- Call Credentials which are attached to a call (or `ClientContext` in C++).
```
Comment (LisaFC) [inline docs/guides/auth.md]: *Call credentials*, which are attached to a call (or `ClientContext` in C++).

**Comment by LisaFC** [comment]
Minor nits, but this looks really good.

**Comment by jtattermusch** [comment]
LGTM as well.

**Comment by jboeuf** [comment]
Thanks much for the great comments and suggestions. Applied them all.

**Pull request merged**